### ISSUE TITLE
How to consume SQLCipher using WPF

### ISSUE THREAD

**Issue by mawarnes** [praeclarum/sqlite-net#544]
Is it possible to use this library in a WPF application for locally encrypted storage?

**Comment by ericsink** [comment]
Yes, it should be.

More specifically, it is possible to use this with a .NET Framework app.  It doesn't matter whether that app also uses WPF or not.

Start here:

https://github.com/ericsink/SQLitePCL.raw/wiki/How-to-use-SQLCipher-with-SQLite-net

One complication:  If your app is WPF-only, then you probably don't actually have a "portable library".  But you'll need one anyway, because sqlite-net-pcl has a dependency on SQLitePCLRaw.bundle_green, and you need to trick it into substituting SQLitePCLRaw.bundle_sqlcipher in its place.

**Comment by mawarnes** [comment]
ah okay, that would explain some fo the strange errors i'm getting. 
I have tried to add it to a normal class library. 
So first off add a PCL to my solution, and add in the references there?
Then add the appropriate reference in my .Net class library?

**Comment by ericsink** [comment]
Yep. Let me know how that goes. If it's problematic, there is another way, but it's a bit tricky as well.

**Comment by mawarnes** [comment]
would I need to add an init call as if it was a cross platform app in the app.xaml.cs?

**Comment by ericsink** [comment]
You mean the batteries Init call? No.

**Comment by mawarnes** [comment]
yes - SQLitePCL.Batteries_V2.Init(); in a normal xamarin project that should go in the app projects....i.e. android/ios/uwp ...is that correct?

In the case of WPF I assume it would go in the app.xaml.cs file at startup

**Comment by ericsink** [comment]
You shouldn't need the batteries init call.

Yes, add the SQLitePCLRaw.bundle_sqlcipher nuget package to the app projects, as explained in the link I gave you.

**Comment by mawarnes** [comment]
what does the init call do?

**Comment by mawarnes** [comment]
so my solution has a class library, which handles all the database calls, and a wpf project.

What functionality should be in the new PCL? 

Where do I create the database? and which project do I add SQLitePCLRaw.bundle_sqlcipher to?

**Comment by mawarnes** [comment]
so when I call code in the PCL from a .net class library i'm getting a file not found exception System.IO.FileNotFoundException: 'Could not load file or assembly 'SQLite-net, Version=1.1.0.0, Culture=neutral, PublicKeyToken=null' or one of its dependencies. The system cannot find the file specified.'

I've added sqlite-net-pcl to the pcl and added SQLitePCLRaw.bundle_sqlcipher to the .net class library

**Comment by mawarnes** [comment]
according to the documentation here
"You will need to add the nuget package to both your PCL project and your platform-dependent project.", but the link you have listed contradicts that and say to not add it to your app project

**Comment by mawarnes** [comment]
after adding the sql-net-pcl package to the 'app' project i'm getting this
: Method 'sqlite3_bind_blob' in type 'SQLitePCL.SQLite3Provider_e_sqlite3' from assembly 'SQLitePCLRaw.provider.e_sqlite3, Version=1.0.0.0, Culture=neutral, PublicKeyToken=9c301db686d0bd12' does not have an implementation.

**Comment by ericsink** [comment]
At this point I'm not sure how best to help you.

Let's try a different approach.  Ignore everything I said above, and do this:

1.  Remove the sqlite-net-pcl nuget package from everywhere.  In this approach, you won't use it.  You don't need a PCL project either.

2.  Instead, add SQLite.cs (from the src directory in this repository) to your "class library, which handles all the database calls"

3.  At the top of that SQLite.cs file, add `#define USE_SQLITEPCL_RAW`.

4.  In that SQLite.cs file, find the line that contains `SQLitePCL.Batteries_V2.Init()` and remove it.

5.  In that same class library, add the nuget package SQLitePCLRaw.core.  Any other nuget packages related to SQLitePCL should be removed from that class library.

6.  In your app project, add the nuget package SQLitePCLRaw.bundle_sqlcipher

7.  And also in the app project, wherever is the appropriate place to add app-level initialization (app.xaml.cs maybe?), add the `SQLitePCL.Batteries_V2.Init()` call.

**Comment by mawarnes** [comment]
very grateful for your help, just making those changes now

**Comment by mawarnes** [comment]
okay that's looking good, so just to recap

first of all I removed the PCL and any references in any of my projects

My solution now contains 3 projects as follows
project.data - .net class library contains data access functions
project.logic - .net class library contains business logic
project.wpf - contains the wpf functionality

I have added the sqlite.cs file, and the SQLitePCLRaw.core to project.data as described
I have added the sqlcipher package to project.wpf where I make the SQLitePCL.Batteries_V2.Init() when the app runs for the first time, then subsequent connections to the database work.

Also, just to be clear, do I need to run the pragmakey query everytime I create a connection.

So good news!! it looks like it's working, I'd really like to understand whats happening here a bit more, but at least I can crack on with my work now. Having this encryption available is fantastic. Thanks for your help
